lecture-notes/2-vcs.md

```diff
@@ -1859,10 +1859,126 @@ Running `git fetch` will update all of your remote tracking branches. For
 example, if someone updated `main` on the remote, you can see their changes
 (update your local `origin/main`) by fetching.
 
-TODO: Can either `git fetch` from or `git push` to a remote. These update
-*remote-tracking branches*, which can then be used to update your local
-branches via `git merge --ff-only` or `git rebase`. Mention `git pull` is a
-shortcut for fetch+merge but can be unintuitive.
+```console?prompt=$
+$ git fetch
+cedar% git fetch
+remote: Enumerating objects: 22, done.
+remote: Counting objects: 100% (22/22), done.
+remote: Compressing objects: 100% (10/10), done.
+remote: Total 16 (delta 12), reused 9 (delta 6), pack-reused 0
+Unpacking objects: 100% (16/16), 7.74 KiB | 792.00 KiB/s, done.
+From github.com:tekknolagi/isdt
+   b357867..d573c7f  mb-vcs-4         -> origin/mb-vcs-4
+ * [new branch]      tom-vcs-lecture2 -> origin/tom-vcs-lecture2
+$
+```
+
+In this output, we can see that the server has printed some messages for us to
+see (the lines starting with `remote:`). Then, Git unpacks the objects---this
+can take a long time in large repositories---and shows a summary of what refs
+have changed. We can see that the remote has updates on its `origin/mb-vcs-4`
+branch and on its (new to us!) `origin/tom-vcs-lecture2` branch.
+
+I use fetch a lot. Right before I go offline---say, for a plane or train
+trip---I'll fetch all of the changes to the upstream repo so that I have them
+available when I'm offline.
+
+Again: fetch *only* updates remote tracking branches. It does **not** update
+normal local branches. My `mb-vcs-4` branch will remain untouched until I
+manually incorporate the changes from upstream using `rebase` or (less
+commonly) `merge`.
+
+#### Rebase
+
+You learned about rebase earlier, but for the sake of completeness and
+repetition in learning, I'll show how I might incorporate the remote changes
+into my local branches using `rebase`. First, we'll take stock of the situation
+and reflexively type `git status` to see what's going on:
+
+```console?prompt=$
+$ git status
+On branch mb-vcs-4
+Your branch is behind 'origin/mb-vcs-4' by 1 commit, and can be fast-forwarded.
+  (use "git pull" to update your local branch)
+$
+```
+
+As expected, the remote has updates that we do not, so our local branch is
+*behind*. The status message also notes that our local branch can be
+*fast-forwarded*, which we talked about earlier as well. This is the easiest
+case and is illustrated by the below diagram:
+
+```
+* a879406 (origin/mb-vcs-4) A shiny new thing, freshly downloaded
+* 5e72155 (mb-vcs-4) My most recent local commit
+* 2187cc4 My previous local commit
+.
+.
+.
+```
+
+It's not the *only* situation we could be in, though. We could also have a
+situation where the remote tracking branch has a new commit **and** the local
+branch has a *different* new commit. In that case, Git will note that the
+branches have "diverged" and will not let you push to the remote branch:
```

Comment:
I think it's a bit jarring to talk about `git push` at all right here, in the middle of a discussion of rebasing. I think it'd be better to talk about fast-forward merges first (`merge --ff-only`), then show how that fails in this situation instead of showing that push fails.

Reply:
This comment still applies. I think talking about fast-forwards (`merge --ff-only`) before talking about rebase makes more sense, since fast-forwards are far simpler.

Reply:
turned into footnote